ui/app/components/totp-edit.js

```diff
@@ -57,29 +64,45 @@ export default class TotpEdit extends Component {
 
   @action
   reset() {
-    this.model.unloadRecord();
-    this.successCallback(null);
+    this.args.model.unloadRecord();
+    this.transitionToRoute(SHOW_ROUTE, this.args.model.name);
   }
 
   @action
-  delete() {
-    this.persist('destroyRecord', () => {
+  async deleteKey() {
+    try {
+      await this.args.model.destroyRecord();
       this.transitionToRoute(LIST_ROOT_ROUTE);
-    });
-  }
-
-  @action
-  create(event) {
-    event.preventDefault();
-    const modelId = this.model.name;
-    // prevent from submitting if there's no key
-    // maybe do something fancier later
-    if (isBlank(modelId)) {
+    } catch (err) {
+      // err will display via model state
       return;
     }
-
-    this.persist('save', () => {
-      this.transitionToRoute(SHOW_ROUTE, modelId);
-    });
+    this.flashMessages.success('Successfully deleted key');
   }
+
+  createKey = task(
+    waitFor(async (event) => {
+      event.preventDefault();
+      const { isValid, state, invalidFormMessage } = this.args.model.validate();
+      this.modelValidations = isValid ? null : state;
+      this.invalidFormAlert = invalidFormMessage;
+
+      if (!isValid) return;
+      try {
+        const allFields = [...this.keyFormFields, ...Object.values(this.groups).flat()];
+        await this.args.model.save({
+          adapterOptions: {
+            keyFormFields: allFields,
+          },
+        });
+        const { generate, exported } = this.args.model;
+        if (generate && exported) this.hasGenerated = true;
+        else this.reset();
```

Comment:
I think we should avoid unloading the record here, there are some challenging bugs to track down because we manually unload records and then try to access properties on them. As I mentioned above there have been some improvements to ember data, so I think manually unloading records is something we can move away from. 

I also think the block conditional here makes it easier to discern these are mutually exclusive
```suggestion
        if (generate && exported) {
        // stay in this template and show QR code returned from response
        this.hasGenerated = true;
        } else {
        // nothing is returned from response, transition to key details route
        this.transitionToRoute(SHOW_ROUTE, this.args.model.name);
        }
```